### ISSUE TITLE
Block editor's apply and cancel buttons gone in visual stepping mode

### ISSUE THREAD

**Issue by ToonTalk** [jmoenig/Snap--Build-Your-Own-Blocks#1853]
When I enable visual stepping I see

![image](https://user-images.githubusercontent.com/2138696/30100335-d45db4c8-92e8-11e7-9f46-9737210b6977.png)

instead of

![image](https://user-images.githubusercontent.com/2138696/30100371-f153e584-92e8-11e7-8918-1ae38ae191af.png)

In release and dev versions.

**Comment by jguille2** [comment]
Hi @ToonTalk,

I think this behavior is necessary.

In VisualStepping, block editor is liveness. You can edit a block that is running and this block changes dynamically. Cancel makes no sense (changes have already been implemented).

**Comment by jmoenig** [comment]
Hi Ken, Joan is right, in "live" mode changes in the the block editor take place immediately, which is both exciting and scary :-)
But - bottom line - this is the intended behavior for visual stepping, which needs to be "live" so we can show which blocks are being evaluated in the block editor.

**Comment by ToonTalk** [comment]
Makes sense.

I wonder if "OK" is the right label since that is normally short for "apply changes and close editor". In stepper mode it means "close editor". I was confused for a while by the current behavior.

**Comment by jmoenig** [comment]
yes, good point, Ken. Let's revisit this!

**Comment by cycomachead** [comment]
CC @brollb Since netsblox works this way all the time.



> I wonder if "OK" is the right label since that is normally short for "apply changes and close editor". In stepper mode it means "close editor". I was confused for a while by the current behavior.

The problem is that the changes are _already_ applied. "Close" would imply "close and don't save" -- which would be wrong under the current paradigm -- but maybe there are other solutions. (We don't really distinguish between Close and Cancel in many places. That could help.)

FWIW, I find the need to apply changes somewhat unnecessary. I'd much rather see a default apply and a "Close and Revert" button that resets a block to a snapshot of when the editor was opened. This is somewhat less granular than than the current "Apply" button, but it would be compatible with stepping mode, and I think is a net-win because more often you do want to keep applying changes -- it would certainly save me lots of clicking!

**Comment by brianharvey** [comment]
I agree that instant-apply would be fine _if_ combined with a revert option.  I wouldn't make it "close and revert"; it's more reassuring if you see just what you reverted.

This would be a huge flag day for teachers, though, so I vote for doing it over xmas break rather than in 4.1.

**Comment by cycomachead** [comment]
That seems fine with me. 

I would suggest and reiterate the need again, that we ought to have buttons in morphic which have a visual disabled state!
(This probably also applies to the original issue)

We should show the buttons in all modes but in when they don’t make sense, the buttons should look washed out / pale and not be clickable. Optionally they’d have a tooltip explaining why they are disabled. 

-- 
Michael Ball
From my iPhone
michaelball.co
